azurerm/internal/services/web/static_site_custom_domain_resource.go

```diff
@@ -0,0 +1,218 @@
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"validation_type": {
+				Type:     pluginsdk.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					txtValidationType,
+					cnameValidationType,
+				}, false),
+			},
+
+			"validation_token": {
+				Type:      pluginsdk.TypeString,
+				Sensitive: true,
+				Computed:  true,
+			},
+		},
+	}
+}
+
+func resourceStaticSiteCustomDomainCreateOrUpdate(d *pluginsdk.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Web.StaticSitesClient
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	log.Printf("[INFO] preparing arguments for AzureRM Static Site custom domain creation.")
+
+	staticSiteId, err := parse.StaticSiteID(d.Get("static_site_id").(string))
+
+	if err != nil {
+		return err
+	}
+
+	domainName := d.Get("name").(string)
+
+	resourceGroup := staticSiteId.ResourceGroup
+	staticSiteName := staticSiteId.Name
+
+	_, err = client.GetStaticSite(ctx, resourceGroup, staticSiteName)
+	if err != nil {
+		return fmt.Errorf("finding Static Site (Name %q / Resource Group %q): %s", staticSiteName, resourceGroup, err)
+	}
+
+	if d.IsNewResource() {
+		existing, err := client.GetStaticSiteCustomDomain(ctx, staticSiteId.ResourceGroup, staticSiteName, domainName)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", staticSiteId, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_static_site_custom_domain", staticSiteId.ID())
+		}
+	}
+
+	validationMethod := d.Get("validation_type").(string)
+
+	siteEnvelope := web.StaticSiteCustomDomainRequestPropertiesARMResource{
+		StaticSiteCustomDomainRequestPropertiesARMResourceProperties: &web.StaticSiteCustomDomainRequestPropertiesARMResourceProperties{
+			ValidationMethod: &validationMethod,
+		},
+	}
+
+	future, err := client.CreateOrUpdateStaticSiteCustomDomain(ctx, staticSiteId.ResourceGroup, staticSiteName, domainName, siteEnvelope)
+	if err != nil {
+		return fmt.Errorf("creating %s: %+v", staticSiteId, err)
+	}
+
+	// we can't wait for the future to be complete for txt validation as we need to give the user the validation token
+	if validationMethod == cnameValidationType {
+		err := future.WaitForCompletionRef(ctx, client.Client)
+		if err != nil {
+			return fmt.Errorf("waiting for creation of Static Site Custom Domain %q (Static Site %q / Resource Group %q): %+v", domainName, staticSiteName, staticSiteId.ResourceGroup, err)
+		}
+	}
+
+	// we need to poll till the validation token is ready
+	if validationMethod == txtValidationType {
+		stateConf := &pluginsdk.StateChangeConf{
+			Pending: []string{
+				string(web.CustomDomainStatusRetrievingValidationToken),
+			},
+			Target: []string{
+				string(web.CustomDomainStatusValidating),
+			},
```

Comment:
This feels very brittle, we're essentially leaving the resource hanging partially complete here. If there's no way to obtain the token in advance of resource creation, we may not be able to support this method of validation in the resource since we can't tell if it's actually successful?
I recognise this is the same behaviour as the portal, but that interactive journey is obviously very different to an automated process.

Reply:
😢 it's a weird one but for our usecase we need to be able to do TXT based validation.